### PR TITLE
Add npm7 support

### PR DESCRIPTION
**Pull request by AlexHladin**
This PR adds support for npm ver. 7.
The npm security audit output  for ver. 6 looks like this:
```json
{
  "actions": [
  ],
  "advisories": {
    "1594": {
      "findings": [
        ...
      ],
      "recommendation": "Upgrade to 0.21.1 or later.",
      "references": "- [Github Issue](https://github.com/axios/axios/issues/3369)\n- [Fix commit](https://github.com/axios/axios/commit/c7329fefc890050edd51e40e469a154d0117fc55)\n- [GitHub Advisory](https://github.com/advisories/GHSA-4w2v-q235-vp99)\n- [Snyk Report](https://snyk.io/vuln/SNYK-JS-AXIOS-1038255)",
      "access": "public",
      "severity": "high",
      "cwe": "CWE-918",
      "metadata": {
        "module_type": "",
        "exploitability": 5,
        "affected_components": ""
      },
      "url": "https://npmjs.com/advisories/1594"
    }
  },
  "mut
```
New version of npm security audit looks like this:
```json
{
  "auditReportVersion": 2,
  "vulnerabilities": {
    "axios": {
      "name": "axios",
      "severity": "high",
      "via": [
        {
          "source": 1594,
          "name": "axios",
          "dependency": "axios",
          "title": "Server-Side Request Forgery",
          "url": "https://npmjs.com/advisories/1594",
          "severity": "high",
          "range": "<0.21.1"
        }
      ]
    },
```

So, the output for the audit command has been changed. Current PR fix incompatibility with npm 7 and safe compatibility with ver. 6.

### CODE REVIEW

File: lib/common.js [hunk 1]
```diff
@@ -78,14 +78,25 @@ function runProgram(command, args, options, stdoutListener, stderrListener) {
...
           stdoutListener(data);
         } catch (error) {
           stderrListener(error);
         }
       })
     );
-  return new Promise((resolve) => {
+  return new Promise((resolve, reject) => {
     proc.on("close", () => resolve());
+    proc.on("error", (error) => reject(error));
```
Comment (quinnturner) [inline lib/common.js:99]: This looks like a nice fix, good find!

File: lib/npm-auditer.js [hunk 2]
```diff
@@ -40,11 +40,13 @@ function printReport(parsedOutput, levels, reportType) {
       printReportObj("NPM audit report JSON:", parsedOutput);
       break;
     case "important": {
-      const relevantAdvisories = Object.keys(parsedOutput.advisories).reduce(
+      const vulnerabilities =
+        parsedOutput.advisories || parsedOutput.vulnerabilities;
+      const relevantAdvisories = Object.keys(vulnerabilities).reduce(
         (acc, advisory) =>
-          levels[parsedOutput.advisories[advisory].severity]
+          levels[vulnerabilities[advisory]]
```
Comment (quinnturner) [inline lib/npm-auditer.js:47]: I think you're missing `.severity` here

File: lib/Model.js [hunk 3]
```diff
@@ -78,7 +84,7 @@ class Model {
     failedLevelsFound.sort();
 
     const advisoriesFound = [
-      ...new Set(this.advisoriesFound.map(advisoryMapper)),
+      ...new Set(this.advisoriesFound.map(advisoryMapper).filter(Boolean)),
```
Comment (quinnturner) [inline lib/Model.js:87]: Out of curiosity, why is this `.filter(Boolean)` necessary?

File: lib/Model.js [hunk 4]
```diff
@@ -44,31 +44,37 @@ class Model {
...
+        return;
+      }
     }
 
     this.advisoriesFound.push(advisory);
   }
 
   load(parsedOutput) {
-    Object.values(parsedOutput.advisories).forEach((a) => this.process(a));
+    if (parsedOutput.advisories) {
```
Comment (quinnturner) [inline lib/Model.js:71]: Can you add a comment here to say NPM 6 (and below add a comment to say NPM 7)

File: lib/Model.js [hunk 5]
```diff
@@ -44,31 +44,37 @@ class Model {
...
+      Object.values(parsedOutput.advisories).forEach((a) => this.process(a));
+      return this.getSummary();
+    }
 
-    return this.getSummary();
+    Object.values(parsedOutput.vulnerabilities).forEach((a) => this.process(a));
+    return this.getSummary((a) => a.via[0].source);
```
Comment (quinnturner) [inline lib/Model.js:77]: Was not expecting this argument to be used other than for tests, glad it did!

File: package-lock.json [hunk 6]
```diff
@@ -1,8 +1,3785 @@
 {
   "name": "audit-ci",
   "version": "3.1.1",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
```
Comment (quinnturner) [inline package-lock.json:4]: This project still technically supports Node 8, which is way behind the Node 15 that NPM 7 is typically used. Will this bring friction to newcomers? It's not documented that this project would use NPM 7. My instinct is that this should stay v1.

**Review by quinnturner — CHANGES_REQUESTED** [review]
Thank you very much for the contribution! I am excited to get it merged. I have laid out a few comments regarding the current implementation.

File: lib/Model.js (hunk 4 above)
Comment (AlexHladin) [reply lib/Model.js:85]: Done

File: package-lock.json (hunk 6 above)
Comment (AlexHladin) [reply package-lock.json]: Reverted

File: lib/Model.js (hunk 3 above)
Comment (AlexHladin) [reply lib/Model.js]: Not really necessary, so I deleted it.

File: lib/npm-auditer.js (hunk 2 above)
Comment (AlexHladin) [reply lib/npm-auditer.js]: Yes, thank you

**Comment by AlexHladin** [comment]
> Thank you very much for the contribution! I am excited to get it merged. I have laid out a few comments regarding the current implementation.
> 
> It would be nice to add NPM 7 to the test suites.

Thank you for your comments, they were very helpful.
I have fixed tests and added tests for npm7. 

Feel free to add any additional suggestions or comments.

File: lib/npm-auditer.js [hunk 7]
```diff
@@ -40,18 +40,32 @@ function printReport(parsedOutput, levels, reportType) {
...
+          (acc, advisory) =>
+            levels[parsedOutput.advisories[advisory].severity]
+              ? {
+                  [advisory]: parsedOutput.advisories[advisory],
+                  ...acc,
+                }
+              : acc,
+          {}
+        );
+      }
```
Comment (quinnturner) [inline lib/npm-auditer.js:68]:
```ts
const advisories = parsedOutput.auditReportVersion === 2 ? parsedOutput.vulnerabilities : parsedOutput.advisories;
const relevantAdvisories = Object.keys(advisories).reduce(
           (acc, advisory) =>
             levels[advisories[advisory].severity]
               ? {
                   [advisory]: parsedOutput.advisories[advisory],
                   ...acc,
                 }
               : acc,
           {}
         );
```

Comment (quinnturner) [reply lib/npm-auditer.js:68]: The PR looks great, maybe this can be simplified? Very very excited to get this in 👍🏻

**Review by quinnturner — CHANGES_REQUESTED** [review]

File: lib/npm-auditer.js (hunk 7 above)
Comment (AlexHladin) [reply lib/npm-auditer.js:68]: Done

**Comment by quinnturner** [comment]
Did a local test, here's some results:

`package.json`:
```json
{
  "name": "audit-ci-npm-high-vulnerability",
  "description": "Test package.json with high vulnerability",
  "dependencies": {
    "cryo": "0.0.6",
    "micromatch": "2.3.0"
  }
}
```

`audit-ci --low`

Result:
```json
audit-ci version: 3.1.1
NPM audit report results:
{
  "advisories": {
    "micromatch": {
      "name": "micromatch",
      "severity": "low",
      "via": [
        "braces"
      ],
      "effects": [],
      "range": "0.2.0 - 2.3.11",
      "nodes": [
        "node_modules/micromatch"
      ],
      "fixAvailable": {
        "name": "micromatch",
        "version": "4.0.2",
        "isSemVerMajor": true
      }
    },
    "cryo": {
      "name": "cryo",
      "severity": "high",
      "via": [
        {
          "source": 690,
          "name": "cryo",
          "dependency": "cryo",
          "title": "Code Injection",
          "url": "https://npmjs.com/advisories/690",
          "severity": "high",
          "range": ">=0.0.0"
        }
      ],
      "effects": [],
      "range": "*",
      "nodes": [
        "node_modules/cryo"
      ],
      "fixAvailable": false
    },
    "braces": {
      "name": "braces",
      "severity": "low",
      "via": [
        {
          "source": 786,
          "name": "braces",
          "dependency": "braces",
          "title": "Regular Expression Denial of Service",
          "url": "https://npmjs.com/advisories/786",
          "severity": "low",
          "range": "<2.3.1"
        }
      ],
      "effects": [
        "micromatch"
      ],
      "range": "<2.3.1",
      "nodes": [
        "node_modules/braces"
      ],
      "fixAvailable": {
        "name": "micromatch",
        "version": "4.0.2",
        "isSemVerMajor": true
      }
    }
  },
  "metadata": {
    "vulnerabilities": {
      "info": 0,
      "low": 2,
      "moderate": 0,
      "high": 1,
      "critical": 0,
      "total": 3
    },
    "dependencies": {
      "prod": 43,
      "dev": 0,
      "optional": 0,
      "peer": 0,
      "peerOptional": 0,
      "total": 42
    }
  }
}
Failed security audit due to high, low vulnerabilities.
Vulnerable advisories are: 786, 690, braces
Exiting...
```

The vulnerable advisories list of `786, 690, braces` does not conform to the current implementation of Yarn and NPM v6's auditing.

To conform to the existing standard, the advisories list is `786` and `690`. `braces` itself is not the advisory, the advisory is `786`.

The current NPM 7 implementation does not meet the standard for specificity around allowlisting. Allowlisting `786` works, but not when provided the allowlist `786|micromatch>braces` (the most precise allowlist for the existing implementation) (the existing approach does not support the allowlist `786|micromatch`, it must be either `786` or `786|micromatch>braces`).

Unfortunately, NPM 7 doesn't give us the paths for convenient usage. I think the solution involves building an allowlisting tree by using `via`, in this case, `"via": [ "braces" ]`. It would have to be recursive to support deeply nested `advisory|path1>path2>path3>path4>...`. I acknowledge that it's more complex to code, but I'd prefer to maintain existing `audit-ci.json` files when transitioning from NPM 6 -> NPM 7, as the underlying advisories have not changed 😄

I think in the `load(parsedOutput)` of `Model.js`, we can build a tree structure and either pass in a format that conforms to the existing `this.process(a)` or create a separate process function (hopefully with some de-duplicated logic).

Let me know if you're still up for doing that! You've done a great job so far 👍🏻

**Comment by AlexHladin** [comment]
> Did a local test, here's some results:
> 
> `package.json`:
> 
> ```json
> {
>   "name": "audit-ci-npm-high-vulnerability",
>   "description": "Test package.json with high vulnerability",
>   "dependencies": {
>     "cryo": "0.0.6",
>     "micromatch": "2.3.0"
>   }
> }
> ```
> 
> `audit-ci --low`
> 
> Result:
> 
> ```json
> audit-ci version: 3.1.1
> NPM audit report results:
> {
>   "advisories": {
>     "micromatch": {
>       "name": "micromatch",
>       "severity": "low",
>       "via": [
>         "braces"
>       ],
>       "effects": [],
>       "range": "0.2.0 - 2.3.11",
>       "nodes": [
>         "node_modules/micromatch"
>       ],
>       "fixAvailable": {
>         "name": "micromatch",
>         "version": "4.0.2",
>         "isSemVerMajor": true
>       }
>     },
>     "cryo": {
>       "name": "cryo",
>       "severity": "high",
>       "via": [
>         {
>           "source": 690,
>           "name": "cryo",
>           "dependency": "cryo",
>           "title": "Code Injection",
>           "url": "https://npmjs.com/advisories/690",
>           "severity": "high",
>           "range": ">=0.0.0"
>         }
>       ],
>       "effects": [],
>       "range": "*",
>       "nodes": [
>         "node_modules/cryo"
>       ],
>       "fixAvailable": false
>     },
>     "braces": {
>       "name": "braces",
>       "severity": "low",
>       "via": [
>         {
>           "source": 786,
>           "name": "braces",
>           "dependency": "braces",
>           "title": "Regular Expression Denial of Service",
>           "url": "https://npmjs.com/advisories/786",
>           "severity": "low",
>           "range": "<2.3.1"
>         }
>       ],
>       "effects": [
>         "micromatch"
>       ],
>       "range": "<2.3.1",
>       "nodes": [
>         "node_modules/braces"
>       ],
>       "fixAvailable": {
>         "name": "micromatch",
>         "version": "4.0.2",
>         "isSemVerMajor": true
>       }
>     }
>   },
>   "metadata": {
>     "vulnerabilities": {
>       "info": 0,
>       "low": 2,
>       "moderate": 0,
>       "high": 1,
>       "critical": 0,
>       "total": 3
>     },
>     "dependencies": {
>       "prod": 43,
>       "dev": 0,
>       "optional": 0,
>       "peer": 0,
>       "peerOptional": 0,
>       "total": 42
>     }
>   }
> }
> Failed security audit due to high, low vulnerabilities.
> Vulnerable advisories are: 786, 690, braces
> Exiting...
> ```
> 
> The vulnerable advisories list of `786, 690, braces` does not conform to the current implementation of Yarn and NPM v6's auditing.
> 
> To conform to the existing standard, the advisories list is `786` and `690`. `braces` itself is not the advisory, the advisory is `786`.
> 
> The current NPM 7 implementation does not meet the standard for specificity around allowlisting. Allowlisting `786` works, but not when provided the allowlist `786|micromatch>braces` (the most precise allowlist for the existing implementation) (the existing approach does not support the allowlist `786|micromatch`, it must be either `786` or `786|micromatch>braces`).
> 
> Unfortunately, NPM 7 doesn't give us the paths for convenient usage. I think the solution involves building an allowlisting tree by using `via`, in this case, `"via": [ "braces" ]`. It would have to be recursive to support deeply nested `advisory|path1>path2>path3>path4>...`. I acknowledge that it's more complex to code, but I'd prefer to maintain existing `audit-ci.json` files when transitioning from NPM 6 -> NPM 7, as the underlying advisories have not changed 😄
> 
> I think in the `load(parsedOutput)` of `Model.js`, we can build a tree structure and either pass in a format that conforms to the existing `this.process(a)` or create a separate process function (hopefully with some de-duplicated logic).
> 
> Let me know if you're still up for doing that! You've done a great job so far 👍🏻

Could you create a few test-cases that describes an issue with building an allowlisting tree? It will be really helpful and I will fix implementation faster.

**Comment by AlexHladin** [comment]
@quinnturner I have updated PR and made tests for audit output for ver.7 equal with tests for ver.6. Please, check the results.

**Comment by quinnturner** [comment]
Hi @AlexHladin! Thanks again **very much** for your updates!

I have merged two PRs, #177 and #178. They may improve (or impact) the success of this PR's pipeline. I suggest you apply `git rebase master` to obtain the CircleCI `package-lock.json` updates.

Secondly, at this time, this project supports Node 8. `Array.prototype.flatMap` was introduced [sometime within Node 11](https://node.green/#ES2019-features-Array-prototype--flat--flatMap-). Considering how useful `flatMap` is in the code that you updated, I am OK with replacing its usage with a polyfill or shim. A quick search resulted in: https://github.com/es-shims/Array.prototype.flatMap. Just be sure to use NPM 6 when applying `npm install array.prototype.flatmap` 😄

**Comment by AlexHladin** [comment]
> Hi @AlexHladin! Thanks again **very much** for your updates!
> 
> I have merged two PRs, #177 and #178. They may improve (or impact) the success of this PR's pipeline. I suggest you apply `git rebase master` to obtain the CircleCI and `package-lock.json` updates.
> 
> Secondly, at this time, this project supports Node 8. `Array.prototype.flatMap` was introduced [sometime within Node 11](https://node.green/#ES2019-features-Array-prototype--flat--flatMap-). Considering how useful `flatMap` is in the code that you updated, I am OK with replacing its usage with a polyfill or shim. A quick search resulted in: https://github.com/es-shims/Array.prototype.flatMap. Just be sure to use NPM 6 when applying `npm install array.prototype.flatmap` 😄

Rebase and `flatMap` fix has been applied successfully 👍

**Comment by quinnturner** [comment]
Thank you!

I found an edge case, and I am not sure how to proceed. I actually think the new behavior is OK, I just want to document it here for future reference. It's related to the `npm-allowlist-path` test. When performing a path allowlist, we need fewer paths to successfully allowlist in NPM 7 compared to NPM 6. Here's the reproduction:

`audit.json` config:

```json
{
  "low": true,
  "allowlist": [
    "880|axios",
    "880|github-build>axios",
    "1594|axios",
    "1594|github-build>axios"
  ]
}
```

`package.json` config:

```json
{
  "name": "audit-ci-npm-allowlisted-path",
  "dependencies": {
    "axios": "^0.15.3",
    "github-build": "^1.2.0"
  }
}
```

<details>
  <summary>When running `audit-ci` in NPM 6, the result is a clean success:</summary>
  
  ```

audit-ci version: 3.1.1
NPM audit report results:
{
  "advisories": {
    "880": {
      "findings": [
        {
          "version": "0.15.3",
          "paths": [
            "axios"
          ]
        },
        {
          "version": "0.15.3",
          "paths": [
            "github-build>axios"
          ]
        }
      ],
      "id": 880,
      "created": "2019-05-06T19:14:45.155Z",
      "updated": "2019-06-03T17:55:10.532Z",
      "deleted": null,
      "title": "Denial of Service",
      "found_by": {
        "link": "https://nornagon.net/",
        "name": "Jeremy Apthorp"
      },
      "reported_by": {
        "link": "https://nornagon.net/",
        "name": "Jeremy Apthorp"
      },
      "module_name": "axios",
      "cves": [],
      "vulnerable_versions": "<0.18.1",
      "patched_versions": ">=0.18.1",
      "overview": "Versions of `axios` prior to 0.18.1 are vulnerable to Denial of Service. If a request exceeds the `maxContentLength` property, the package prints an error but does not stop the request. This may cause high CPU usage and lead to Denial of Service.",
      "recommendation": "Upgrade to 0.18.1 or later.",
      "references": "- [Github Issue](https://github.com/axios/axios/issues/1098)\n- [Snyk Report](https://snyk.io/vuln/SNYK-JS-AXIOS-174505)",
      "access": "public",
      "severity": "moderate",
      "cwe": "CWE-400",
      "metadata": {
        "module_type": "",
        "exploitability": 4,
        "affected_components": ""
      },
      "url": "https://npmjs.com/advisories/880"
    },
    "1594": {
      "findings": [
        {
          "version": "0.15.3",
          "paths": [
            "axios"
          ]
        },
        {
          "version": "0.15.3",
          "paths": [
            "github-build>axios"
          ]
        }
      ],
      "id": 1594,
      "created": "2021-01-04T21:04:59.346Z",
      "updated": "2021-01-04T21:05:54.214Z",
      "deleted": null,
      "title": "Server-Side Request Forgery",
      "found_by": {
        "link": "",
        "name": "Anonymous",
        "email": ""
      },
      "reported_by": {
        "link": "",
        "name": "Anonymous",
        "email": ""
      },
      "module_name": "axios",
      "cves": [
        "CVE-2020-28168"
      ],
      "vulnerable_versions": "<0.21.1",
      "patched_versions": ">=0.21.1",
      "overview": "The `axios` NPM package before 0.21.1 contains a Server-Side Request Forgery (SSRF) vulnerability where an attacker is able to bypass a proxy by providing a URL that responds with a redirect to a restricted host or IP address.",
      "recommendation": "Upgrade to 0.21.1 or later.",
      "references": "- [Github Issue](https://github.com/axios/axios/issues/3369)\n- [Fix commit](https://github.com/axios/axios/commit/c7329fefc890050edd51e40e469a154d0117fc55)\n- [GitHub Advisory](https://github.com/advisories/GHSA-4w2v-q235-vp99)\n- [Snyk Report](https://snyk.io/vuln/SNYK-JS-AXIOS-1038255)",
      "access": "public",
      "severity": "high",
      "cwe": "CWE-918",
      "metadata": {
        "module_type": "",
        "exploitability": 5,
        "affected_components": ""
      },
      "url": "https://npmjs.com/advisories/1594"
    }
  },
  "metadata": {
    "vulnerabilities": {
      "info": 0,
      "low": 0,
      "moderate": 2,
      "high": 2,
      "critical": 0
    },
    "dependencies": 8,
    "devDependencies": 0,
    "optionalDependencies": 0,
    "totalDependencies": 8
  }
}
Passed npm security audit.  
  ```
</details>



<details>
  <summary>When running with NPM 7, it's a success but with the warning "Consider not allowlisting path: 1594|github-build>axios."</summary>
  
  ```
  audit-ci version: 3.1.1
NPM audit report results:
{
  "advisories": {
    "github-build": {
      "name": "github-build",
      "severity": "moderate",
      "via": [
        "axios"
      ],
      "effects": [],
      "range": "<=1.2.0",
      "nodes": [
        "node_modules/github-build"
      ],
      "fixAvailable": true
    },
    "axios": {
      "name": "axios",
      "severity": "high",
      "via": [
        {
          "source": 1594,
          "name": "axios",
          "dependency": "axios",
          "title": "Server-Side Request Forgery",
          "url": "https://npmjs.com/advisories/1594",
          "severity": "high",
          "range": "<0.21.1"
        },
        {
          "source": 880,
          "name": "axios",
          "dependency": "axios",
          "title": "Denial of Service",
          "url": "https://npmjs.com/advisories/880",
          "severity": "moderate",
          "range": "<0.18.1"
        }
      ],
      "effects": [
        "github-build"
      ],
      "range": "<=0.21.0",
      "nodes": [
        "node_modules/axios",
        "node_modules/github-build/node_modules/axios"
      ],
      "fixAvailable": {
        "name": "axios",
        "version": "0.21.1",
        "isSemVerMajor": true
      }
    }
  },
  "metadata": {
    "vulnerabilities": {
      "info": 0,
      "low": 0,
      "moderate": 1,
      "high": 1,
      "critical": 0,
      "total": 2
    },
    "dependencies": {
      "prod": 9,
      "dev": 0,
      "optional": 0,
      "peer": 0,
      "peerOptional": 0,
      "total": 8
    }
  }
}
Consider not allowlisting path: 1594|github-build>axios.
Passed npm security audit.
  ```
 </details>
 
 I think this is OK. I _think_ we do not receive as much information in the NPM 7 audit in this case. Also, we don't need the config to be overly (and unnecessarily) verbose. I think I am OK with it as-is, but I'd like to hear your thoughts before I merge!

**Comment by AlexHladin** [comment]
> Thank you!
> 
> I found an edge case, and I am not sure how to proceed. I actually think the new behavior is OK, I just want to document it here for future reference. It's related to the `npm-allowlist-path` test. When performing a path allowlist, we need fewer paths to successfully allowlist in NPM 7 compared to NPM 6. Here's the reproduction:
> 
> `audit.json` config:
> 
> ```json
> {
>   "low": true,
>   "allowlist": [
>     "880|axios",
>     "880|github-build>axios",
>     "1594|axios",
>     "1594|github-build>axios"
>   ]
> }
> ```
> 
> `package.json` config:
> 
> ```json
> {
>   "name": "audit-ci-npm-allowlisted-path",
>   "dependencies": {
>     "axios": "^0.15.3",
>     "github-build": "^1.2.0"
>   }
> }
> ```
> 
> When running `audit-ci` in NPM 6, the result is a clean success:
> When running with NPM 7, it's a success but with the warning "Consider not allowlisting path: 1594|github-build>axios."
> I think this is OK. I _think_ we do not receive as much information in the NPM 7 audit in this case. Also, we don't need the config to be overly (and unnecessarily) verbose. I think I am OK with it as-is, but I'd like to hear your thoughts before I merge!

The current audit output of npm 7 is not so verbose as the output of npm 6. It is a real pain to build a similar path as npm 6 produces. I can try to cover the case above in the test but I am not sure that it will bring any good results or at least the output will be the same.

**Comment by quinnturner** [comment]
Maybe we can document the edge case I listed with a test with NPM 6 and 7. Even though the allowlists are different, as long as the behaviour is covered with a test, we can refer to that new behaviour as being accepted. I think you're right in that the added complexity is probably not worth the effort.

**Comment by AlexHladin** [comment]
> Maybe we can document the edge case I listed with a test with NPM 6 and 7. Even though the allowlists are different, as long as the behaviour is covered with a test, we can refer to that new behaviour as being accepted. I think you're right in that the added complexity is probably not worth the effort.

Could you suggest where I should add this documentation?

**Review by quinnturner — APPROVED** [review]
